### PR TITLE
add a fn-def type and adjust type of fn constants

### PR DESCRIPTION
**Pull request by nikomatsakis**
but we're not really testing it in any useful way.

cc @voidc

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Huh, tests were passing locally for me, I thought.

File: racket-src/body/type-of.rkt [hunk 1]
```diff
@@ -125,14 +125,35 @@
...
+
+   ; Split the generics from the function into early/late
+   (where/error (KindedVarId_early ..._n KindedVarId_late ...) KindedVarIds)
+
+   ; Create a substitution from early-bound parameters to the values from user
+   (where/error Substitution_early (create-substitution (KindedVarId_early ...) (Parameter ...)))
+   (where/error Biformulas_early (apply-substitution Substitution_early Biformulas))
+
+   ; Apply that substitution
+   (where/error Ty_fn (rigid-ty (fn-def FnId) (Parameter ... KindedVarId_late ...)))
```
Comment (voidc) [inline racket-src/body/type-of.rkt]: I think we need to extract the `VarId` from the `KindedVarId_late` here

Comment (voidc) [reply racket-src/body/type-of.rkt]: I fixed this here: https://github.com/nikomatsakis/a-mir-formality/compare/ty-fn-def...voidc:a-mir-formality:ty-fn-def

Comment (nikomatsakis) [reply racket-src/body/type-of.rkt]: oh, yeah, that's correct

**Comment by nikomatsakis** [comment]
I don't understand why CI is failing :(

**Comment by nikomatsakis** [comment]
Ah, but it actually *started* failing when I merged #95, it appears. Curious.

**Comment by nikomatsakis** [comment]
The PR itself seemed fine, but browsing the actions, I see this: https://github.com/nikomatsakis/a-mir-formality/actions/runs/3054378958/jobs/4926199312

**Comment by nikomatsakis** [comment]
It works locally, I'm going to merge, but we have to figure this out at some point!

**Pull request merged**